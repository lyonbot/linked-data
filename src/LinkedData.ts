import { Schema, SchemaContext, SchemaDescriptor } from './schema';
import { isObject, mapValues } from './utils';
import { getNormalObjectProxy } from './normalObjectProxy';
import { toRef, isDataNodeRef, DataNodeRef, $getDataNode } from './DataNodeRef';
import { transformRaw } from './transformRaw';
import { VOID_NODE_ERROR, warnIfNotPlainObject } from './warnings';

/**
 * @public
 */
export interface LinkedDataOptions {
  schemas?: Record<string, SchemaDescriptor | string> | SchemaContext;
}

/**
 * @public
 */
export interface LinkedDataImportOptions {
  /**
   * while importing, if source data indicated a id, but it is taken by a existing DataNode, shall we rewrite it?
   *
   * - `never` - (default) do not overwrite. a new id will be generated.
   * - `same-schema` - overwrite if the source data has the same schema, otherwise make a new id.
   * - `always` - (dangerous) always overwrite regardless of the source data's schema
   * - a function to decide whether to overwrite. if return `false`, a new id will be generated.
   */
  overwrite?: 'same-schema' | 'always' | 'never' | ((data: any, existingNode: DataNode, schema: Schema) => boolean);
}

/**
 * All you need is this node manager.
 *
 * @public
 */
export class LinkedData {
  schemas: SchemaContext;
  _nodes: Map<string, DataNode> = new Map();

  constructor(options: LinkedDataOptions) {
    const schemas = options.schemas;
    this.schemas = schemas && schemas instanceof SchemaContext ? schemas : new SchemaContext(schemas);
  }

  /**
   * create a DataNode from JSON value.
   *
   * the `value` can NOT be circular!
   *
   * @public
   * @param value - data to be imported
   * @param schema - the schema name of current value
   * @returns
   */
  import<T = any>(value: T, schema?: string | Schema | null, options?: LinkedDataImportOptions) {
    const actualSchema = this.schemas.get(schema);

    const existingNode = toRef(value)?.node;
    if (existingNode) {
      if (existingNode.schema !== actualSchema) throw new Error('Cannot import existing Node with different Schema');
      return existingNode as DataNode<T>; // TODO: support multiple owners
    }

    // the id from raw value, which might duplicate
    const presetId = actualSchema && actualSchema.type === 'object' && actualSchema.readKey(value);
    let id = presetId && presetId !== 0 ? String(presetId) : '';

    const sameIdNode = id && this._nodes.get(id);
    if (sameIdNode) {
      let overwrite = options?.overwrite || 'never';

      if (overwrite === 'same-schema') {
        overwrite = sameIdNode.schema === actualSchema ? 'always' : 'never';
      }

      if (typeof overwrite === 'function') {
        // sameIdNode is not-falsy, implicitly means actualSchema is not-falsy
        overwrite = overwrite(value, sameIdNode, actualSchema!) ? 'always' : 'never';
      }

      if (overwrite === 'never') id = '';
    }

    if (!id) id = this.allocateId(presetId ?? (actualSchema && actualSchema.$id));

    const node = new DataNode(this, id, actualSchema);
    this._nodes.set(id, node);
    node.setValue(value, options);
    return node as DataNode<T>;
  }

  createVoidNode<T = any>(options: { id?: string; schema?: string | Schema | null }) {
    const actualSchema = this.schemas.get(options.schema);
    const id = this.allocateId(options.id ?? (actualSchema && actualSchema.$id));

    const node = new DataNode(this, id, actualSchema);
    this._nodes.set(id, node);
    return node as DataNode<T>;
  }

  getNode(id: string) {
    return this._nodes.get(id);
  }

  allocateId(_ref?: any) {
    const prefix = (typeof _ref === 'string' ? _ref : typeof _ref === 'number' ? String(_ref) : null) || '_node';

    let currentId = prefix;
    for (let index = 1; this._nodes.has(currentId); index++, currentId = prefix + index);
    return currentId;
  }
}

/**
 * @public
 */
export const enum DataNodeStatus {
  VOID,
  FILLED_WITH_ANY,
  FILLED_WITH_REF,
  FILLED_WITH_OBJECT,
  FILLED_WITH_ARRAY,
}

const $SET_TO_VOID = Symbol('$SET_TO_VOID');

/**
 * @public
 */
export class DataNode<T = any> {
  readonly id: string;
  readonly owner: LinkedData;
  readonly schema: Schema | null;

  status: DataNodeStatus = DataNodeStatus.VOID;

  /**
   * the raw value, which might be a primitive, a `DataNodeRef` or a object containing `DataNodeRef` inside.
   *
   * this will NOT contain `DataNode` or Proxy from DataNode
   */
  raw?: any;

  /** the proxy. exists when status is FILLED_WITH_OBJECT or FILLED_WITH_ARRAY */
  _proxy?: any;

  /** if you want to refer this node, use this anywhere you need */
  readonly ref: DataNodeRef;

  constructor(owner: LinkedData, id: string, schema: Schema | null) {
    this.owner = owner;
    this.id = id;
    this.schema = schema;
    this.ref = new DataNodeRef(owner, id);
  }

  /**
   * @internal
   */
  get [$getDataNode]() {
    return this;
  }

  /**
   * export a plain JSON object / array, which might contains circular reference.
   */
  export(options?: {
    /**
     * optional, default: false
     *
     * if true, the output objects will have extra `_id` (or other name depend on schema) property
     */
    writeKey?: boolean | ((node: DataNode) => boolean);
  }): T {
    const writeKey = options && options.writeKey;

    return transformRaw(this.ref, {
      createDest(raw, node) {
        if (node && node.status === DataNodeStatus.VOID) {
          throw new Error(VOID_NODE_ERROR);
        }

        if (!isObject(raw)) return raw;
        if (Array.isArray(raw)) return new Array(raw.length);
        return {};
      },
      fillDest({ dest, recipe, node }) {
        Object.assign(dest, recipe);

        if (
          node &&
          node.schema &&
          node.schema.type === 'object' &&
          writeKey &&
          (typeof writeKey !== 'function' || writeKey(node))
        ) {
          node.schema.writeKey(dest, node.id);
        }
      },
    });
  }

  /**
   * iterate and visit every node.
   *
   * @param visitor
   */
  iterate(visitor: (node: DataNode, path: (string | number)[]) => void | 'abort' | 'skip') {
    const $abort = Symbol('abort');
    try {
      transformRaw(this.ref, {
        createDest(raw, node, path) {
          if (!node) return raw;

          const ans = visitor(node, path);

          if (ans === 'abort') throw $abort;
          if (ans === 'skip') return transformRaw.final(null);

          return raw;
        },
        fillDest: () => void 0,
      });
    } catch (e) {
      if (e !== $abort) throw e;
    }
  }

  /**
   * the proxied current value. you can mutate it directly.
   */
  get value(): T {
    if (this.status === DataNodeStatus.VOID) throw new Error(VOID_NODE_ERROR);
    if (this.status === DataNodeStatus.FILLED_WITH_ANY) return this.raw;
    if (this.status === DataNodeStatus.FILLED_WITH_REF) return (this.raw as DataNodeRef).node.value;

    return this._proxy || (this._proxy = makeProxyForNode(this));
  }

  set value(v: T) {
    this.setValue(v);
  }

  /**
   * set the value
   *
   * @param importOptions - optional, see LinkedData.import
   */
  setValue(value: T, importOptions: LinkedDataImportOptions = {}) {
    this._proxy = void 0;

    // special mark
    if ((value as unknown) === $SET_TO_VOID) {
      this.raw = void 0;
      this.status = DataNodeStatus.VOID;
      return;
    }

    // if not object, just store as-is

    if (!isObject(value)) {
      this.raw = value;
      this.status = DataNodeStatus.FILLED_WITH_ANY;
      return;
    }

    // maybe is a DataNodeRef
    // TODO: add warn

    const ref = toRef(value);
    if (ref) {
      this.raw = ref;
      this.status = DataNodeStatus.FILLED_WITH_REF;
      return;
    }

    // now `value` is an object / array
    // deeply process properties / items

    warnIfNotPlainObject(value);

    this.status = Array.isArray(value) ? DataNodeStatus.FILLED_WITH_ARRAY : DataNodeStatus.FILLED_WITH_OBJECT;
    this.raw = mapValues(value, (v, k) => convertBeforeWriteField(this, v, k, importOptions));
  }

  /**
   * clear node value and set status to void
   */
  setVoid() {
    this.setValue($SET_TO_VOID as any);
  }
}

/**
 * deep clone a value, transform all `DataNodeRef`, `DataNode` and Proxy to `DataNodeRef`
 *
 * @internal
 */
export function deeplyCloneAndNormalizeRefs(x: any) {
  if (!isObject(x)) return x;

  const ref = toRef(x);
  if (ref) return ref;

  warnIfNotPlainObject(x);
  return mapValues(x, deeplyCloneAndNormalizeRefs);
}

function makeProxyForNode(node: DataNode) {
  return new Proxy(node.raw, {
    get: (target, key) => {
      if (key === $getDataNode) return node;

      const v = Reflect.get(target, key);
      if (!isObject(v)) return v;
      if (isDataNodeRef(v)) return v.node?.value;
      return getNormalObjectProxy(v);
    },
    set: (target, key, value) => {
      /* istanbul ignore next */
      if (typeof key === 'symbol') throw new Error(`Do not use symbol property`);

      target[key] = convertBeforeWriteField(node, value, key);
      return true;
    },
  });
}

function convertBeforeWriteField(node: DataNode, value: any, key: any, importOptions?: LinkedDataImportOptions): any {
  const propSchema = node.schema?.get(key);
  if (!propSchema) return deeplyCloneAndNormalizeRefs(value); // no need to create another Node

  const ref = toRef(value);
  if (ref && ref.node.schema !== propSchema) throw new Error('Cannot refer Node whose schema mismatches');

  return ref || node.owner.import(value, propSchema, importOptions).ref;
}
