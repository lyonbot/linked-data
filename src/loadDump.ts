import type { LinkedData, DataNode } from './LinkedData';
import { transformRaw } from './transformRaw';

export interface DumpedNode {
  id: string;
  schemaId: string | null;
  raw: any;
  refs: {
    path: (string | number)[];
    targetNodeId: string;
  }[];
  referredBy: {
    sourceNodeId: string;
    path: (string | number)[];
  }[];
}

const refPlaceholderPrefix = '[[[ ref ]]]';

export function dumpDataNodes(
  node: DataNode,
  options?: {
    writeKey?: boolean;
  },
) {
  const makeInfoWithoutRaw = (node: DataNode): DumpedNode => {
    const schemaId = node.schema?.$id || null;
    if (schemaId) schemasVisited.add(schemaId);

    return {
      id: node.id,
      schemaId,
      refs: [],
      referredBy: [],
      raw: null,
    };
  };

  const writeKey = options?.writeKey;

  const schemasVisited: Set<string> = new Set();
  const nodesToVisit: DataNode[] = [];
  const nodesVisited: Record<string, DumpedNode> = Object.create(null);

  nodesToVisit.push(node);
  nodesVisited[node.id] = makeInfoWithoutRaw(node);

  while (nodesToVisit.length) {
    const node = nodesToVisit.shift()!;
    const visitInfo = nodesVisited[node.id]!;

    visitInfo.raw = transformRaw(node.raw, {
      createDest(raw, toNode, path) {
        if (toNode) {
          let targetNodeInfo = nodesVisited[toNode.id];
          if (!targetNodeInfo) {
            nodesVisited[toNode.id] = targetNodeInfo = makeInfoWithoutRaw(toNode);
            nodesToVisit.push(toNode);
          }

          visitInfo.refs.push({ path, targetNodeId: toNode.id });
          targetNodeInfo.referredBy.push({ sourceNodeId: node.id, path });

          return `${refPlaceholderPrefix} #${visitInfo.refs.length - 1} -> ${toNode.id}`;
          // return transformRaw.final({ $type: 'ref', targetNodeId: toNode.id });
        }

        // now just normal object / array
        if (typeof raw !== 'object' || raw === null) return raw;
        if (Array.isArray(raw)) return new Array(raw.length);
        return {};
      },
      fillDest({ dest, recipe }) {
        Object.assign(dest, recipe);
      },
    });

    if (writeKey && node.schema?.type === 'object') {
      node.schema.writeKey(visitInfo.raw, node.id);
    }
  }

  return {
    schemaIds: Array.from(schemasVisited),
    nodeInfos: nodesVisited,
  };
}

export function loadDataNodes(
  destination: LinkedData,
  nodeInfos: Iterable<DumpedNode> | Record<string, DumpedNode>,
): Record<string, DataNode> {
  const infoArr: DumpedNode[] =
    Symbol.iterator in nodeInfos
      ? Array.from(nodeInfos as Iterable<DumpedNode>) // array or set
      : Object.values(nodeInfos);

  // ----------------------------------------------------------------

  const inflated: Record<string, DataNode> = Object.create(null);

  infoArr.forEach(nodeInfo => {
    inflated[nodeInfo.id] = destination.createVoidNode({
      id: nodeInfo.id,
      schema: nodeInfo.schemaId,
    });
  });

  infoArr.forEach(nodeInfo => {
    const pendingLinks: Record<string, DataNode> = {};
    nodeInfo.refs.forEach(({ path, targetNodeId }) => {
      pendingLinks[JSON.stringify(path)] = inflated[targetNodeId]!;
    });

    inflated[nodeInfo.id].setValue(
      transformRaw(nodeInfo.raw, {
        createDest(raw, _, path) {
          {
            const toNode =
              typeof raw === 'string' && // placeholder is always string
              raw.startsWith(refPlaceholderPrefix) &&
              pendingLinks[JSON.stringify(path)];

            if (toNode) {
              // make a link to something else
              return transformRaw.final(toNode.ref);
            }
          }

          // now just normal object / array
          if (typeof raw !== 'object' || raw === null) return raw;
          if (Array.isArray(raw)) return new Array(raw.length);
          return {};
        },
        fillDest: ({ dest, recipe }) => {
          Object.assign(dest, recipe);
        },
      }),
    );
  });

  return inflated;
}
