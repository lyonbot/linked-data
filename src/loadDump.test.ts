import { dumpDataNodes, loadDataNodes } from './loadDump';
import { LinkedData } from './LinkedData';

describe('loadDump', () => {
  it('should work', () => {
    const ld = new LinkedData({
      schemas: {
        Message: {
          type: 'object',
          key: 'id',
          properties: {
            replies: { type: 'array', items: 'Message' },
          },
        },
      },
    });

    const data = {
      id: 'foo',
      text: 'Hello',
      replies: [
        { id: 'bar', text: 'Hi' },
        { id: 'baz', text: 'Bye' },
        { id: 'foo', text: 'Hey?' }, // duplicated id
      ],
    };

    const node = ld.import(data, 'Message');
    const node2 = ld.createVoidNode({});
    node2.value = node.ref;
    node.value.replies.push(node2.value); // self circle

    const dumped0 = dumpDataNodes(node);
    expect(dumped0).toMatchSnapshot('dumped');

    const dumped = dumpDataNodes(node, { writeKey: true });
    expect(dumped).toMatchSnapshot('dumped with writeKey');
    expect(dumped.nodeInfos['foo.1'].raw.id).not.toBe('foo');
    // console.log(JSON.stringify(dumped, null, 2));

    // ----------------------------------------------------------------

    const loaded = loadDataNodes(ld, dumped.nodeInfos);
    expect(Object.keys(loaded).length).toBe(Object.keys(dumped.nodeInfos).length);

    const node3 = loaded[node.id];
    const dump3 = dumpDataNodes(node3);
    expect(dump3).toMatchSnapshot('loaded');
    expect(node3.id).not.toBe(node.id);
  });
});
