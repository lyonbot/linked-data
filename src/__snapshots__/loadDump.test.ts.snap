// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadDump should work: dumped 1`] = `
Object {
  "nodeInfos": Object {
    "_node": Object {
      "id": "_node",
      "raw": Array [
        "[[[ ref ]]] #0 -> bar",
        "[[[ ref ]]] #1 -> baz",
        "[[[ ref ]]] #2 -> foo2",
        "[[[ ref ]]] #3 -> foo",
      ],
      "referredBy": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "sourceNodeId": "foo",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            0,
          ],
          "targetNodeId": "bar",
        },
        Object {
          "path": Array [
            1,
          ],
          "targetNodeId": "baz",
        },
        Object {
          "path": Array [
            2,
          ],
          "targetNodeId": "foo2",
        },
        Object {
          "path": Array [
            3,
          ],
          "targetNodeId": "foo",
        },
      ],
      "schemaId": "#/Message/replies",
    },
    "bar": Object {
      "id": "bar",
      "raw": Object {
        "id": "bar",
        "text": "Hi",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            0,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "baz": Object {
      "id": "baz",
      "raw": Object {
        "id": "baz",
        "text": "Bye",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            1,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "foo": Object {
      "id": "foo",
      "raw": Object {
        "id": "foo",
        "replies": "[[[ ref ]]] #0 -> _node",
        "text": "Hello",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            3,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "targetNodeId": "_node",
        },
      ],
      "schemaId": "Message",
    },
    "foo2": Object {
      "id": "foo2",
      "raw": Object {
        "id": "foo",
        "text": "Hey?",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            2,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
  },
  "schemaIds": Array [
    "Message",
    "#/Message/replies",
  ],
}
`;

exports[`loadDump should work: dumped with writeKey 1`] = `
Object {
  "nodeInfos": Object {
    "_node": Object {
      "id": "_node",
      "raw": Array [
        "[[[ ref ]]] #0 -> bar",
        "[[[ ref ]]] #1 -> baz",
        "[[[ ref ]]] #2 -> foo2",
        "[[[ ref ]]] #3 -> foo",
      ],
      "referredBy": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "sourceNodeId": "foo",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            0,
          ],
          "targetNodeId": "bar",
        },
        Object {
          "path": Array [
            1,
          ],
          "targetNodeId": "baz",
        },
        Object {
          "path": Array [
            2,
          ],
          "targetNodeId": "foo2",
        },
        Object {
          "path": Array [
            3,
          ],
          "targetNodeId": "foo",
        },
      ],
      "schemaId": "#/Message/replies",
    },
    "bar": Object {
      "id": "bar",
      "raw": Object {
        "id": "bar",
        "text": "Hi",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            0,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "baz": Object {
      "id": "baz",
      "raw": Object {
        "id": "baz",
        "text": "Bye",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            1,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "foo": Object {
      "id": "foo",
      "raw": Object {
        "id": "foo",
        "replies": "[[[ ref ]]] #0 -> _node",
        "text": "Hello",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            3,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "targetNodeId": "_node",
        },
      ],
      "schemaId": "Message",
    },
    "foo2": Object {
      "id": "foo2",
      "raw": Object {
        "id": "foo2",
        "text": "Hey?",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            2,
          ],
          "sourceNodeId": "_node",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
  },
  "schemaIds": Array [
    "Message",
    "#/Message/replies",
  ],
}
`;

exports[`loadDump should work: loaded 1`] = `
Object {
  "nodeInfos": Object {
    "_node3": Object {
      "id": "_node3",
      "raw": Array [
        "[[[ ref ]]] #0 -> bar2",
        "[[[ ref ]]] #1 -> baz2",
        "[[[ ref ]]] #2 -> foo22",
        "[[[ ref ]]] #3 -> foo3",
      ],
      "referredBy": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "sourceNodeId": "foo3",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            0,
          ],
          "targetNodeId": "bar2",
        },
        Object {
          "path": Array [
            1,
          ],
          "targetNodeId": "baz2",
        },
        Object {
          "path": Array [
            2,
          ],
          "targetNodeId": "foo22",
        },
        Object {
          "path": Array [
            3,
          ],
          "targetNodeId": "foo3",
        },
      ],
      "schemaId": "#/Message/replies",
    },
    "bar2": Object {
      "id": "bar2",
      "raw": Object {
        "id": "bar",
        "text": "Hi",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            0,
          ],
          "sourceNodeId": "_node3",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "baz2": Object {
      "id": "baz2",
      "raw": Object {
        "id": "baz",
        "text": "Bye",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            1,
          ],
          "sourceNodeId": "_node3",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "foo22": Object {
      "id": "foo22",
      "raw": Object {
        "id": "foo2",
        "text": "Hey?",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            2,
          ],
          "sourceNodeId": "_node3",
        },
      ],
      "refs": Array [],
      "schemaId": "Message",
    },
    "foo3": Object {
      "id": "foo3",
      "raw": Object {
        "id": "foo",
        "replies": "[[[ ref ]]] #0 -> _node3",
        "text": "Hello",
      },
      "referredBy": Array [
        Object {
          "path": Array [
            3,
          ],
          "sourceNodeId": "_node3",
        },
      ],
      "refs": Array [
        Object {
          "path": Array [
            "replies",
          ],
          "targetNodeId": "_node3",
        },
      ],
      "schemaId": "Message",
    },
  },
  "schemaIds": Array [
    "Message",
    "#/Message/replies",
  ],
}
`;
